File: server/lib/mcp-client.ts
import WebSocket from 'ws';
import { EventEmitter } from 'events';

export interface MCPServer {
  id: string;
  name: string;
  url: string;
  status: 'connected' | 'disconnected' | 'error';
  capabilities?: string[];
  tools?: MCPTool[];
}

export interface MCPTool {
  name: string;
  description: string;
  parameters: Record<string, any>;
}

export interface MCPMessage {
  id: string;
  method: string;
  params?: any;
  result?: any;
  error?: any;
}

export class MCPClient extends EventEmitter {
  private ws: WebSocket | null = null;
  private messageId = 0;
  private pendingRequests = new Map<string, { resolve: Function; reject: Function }>();
  
  constructor(
    public readonly server: MCPServer
  ) {
    super();
  }

  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.server.url);
        
        this.ws.on('open', () => {
          this.server.status = 'connected';
          this.emit('connected', this.server);
          this.initialize().then(resolve).catch(reject);
        });

        this.ws.on('message', (data: string) => {
          try {
            const message: MCPMessage = JSON.parse(data);
            this.handleMessage(message);
          } catch (error) {
            console.error('Failed to parse MCP message:', error);
          }
        });

        this.ws.on('error', (error) => {
          console.error('MCP WebSocket error:', error);
          this.server.status = 'error';
          this.emit('error', error);
          reject(error);
        });

        this.ws.on('close', () => {
          this.server.status = 'disconnected';
          this.emit('disconnected', this.server);
        });

        // Connection timeout
        setTimeout(() => {
          if (this.server.status !== 'connected') {
            reject(new Error('MCP connection timeout'));
          }
        }, 10000);

      } catch (error) {
        this.server.status = 'error';
        reject(error);
      }
    });
  }

  private async initialize(): Promise<void> {
    // Send initialization request
    const response = await this.sendRequest('initialize', {
      protocolVersion: '2024-11-05',
      capabilities: {
        roots: { listChanged: true },
        sampling: {}
      },
      clientInfo: {
        name: 'VoiceAgent-MCP-Client',
        version: '1.0.0'
      }
    });

    if (response.capabilities) {
      this.server.capabilities = Object.keys(response.capabilities);
    }

    // Get available tools
    try {
      const toolsResponse = await this.sendRequest('tools/list', {});
      if (toolsResponse.tools) {
        this.server.tools = toolsResponse.tools;
      }
    } catch (error) {
      console.warn('Failed to get MCP tools:', error);
    }

    // Send initialized notification
    this.sendNotification('notifications/initialized', {});
  }

  async sendRequest(method: string, params?: any): Promise<any> {
    return new Promise((resolve, reject) => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        reject(new Error('MCP connection not open'));
        return;
      }

      const id = (++this.messageId).toString();
      const message: MCPMessage = { id, method, params };

      this.pendingRequests.set(id, { resolve, reject });

      try {
        this.ws.send(JSON.stringify(message));
        
        // Request timeout
        setTimeout(() => {
          if (this.pendingRequests.has(id)) {
            this.pendingRequests.delete(id);
            reject(new Error(`MCP request timeout: ${method}`));
          }
        }, 30000);

      } catch (error) {
        this.pendingRequests.delete(id);
        reject(error);
      }
    });
  }

  sendNotification(method: string, params?: any): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      return;
    }

    const message = { method, params };
    try {
      this.ws.send(JSON.stringify(message));
    } catch (error) {
      console.error('Failed to send MCP notification:', error);
    }
  }

  private handleMessage(message: MCPMessage): void {
    if (message.id && this.pendingRequests.has(message.id)) {
      const { resolve, reject } = this.pendingRequests.get(message.id)!;
      this.pendingRequests.delete(message.id);

      if (message.error) {
        reject(new Error(message.error.message || 'MCP request failed'));
      } else {
        resolve(message.result || {});
      }
    } else {
      // Handle notifications and other messages
      this.emit('message', message);
    }
  }

  async callTool(name: string, arguments_: Record<string, any>): Promise<any> {
    return this.sendRequest('tools/call', {
      name,
      arguments: arguments_
    });
  }

  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.server.status = 'disconnected';
    this.pendingRequests.clear();
  }

  isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }
}

export class MCPManager {
  private clients = new Map<string, MCPClient>();
  private servers = new Map<string, MCPServer>();

  async addServer(server: Omit<MCPServer, 'id' | 'status'>): Promise<MCPServer> {
    const id = `mcp_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    const mcpServer: MCPServer = {
      ...server,
      id,
      status: 'disconnected'
    };

    this.servers.set(id, mcpServer);
    return mcpServer;
  }

  async connectServer(serverId: string): Promise<void> {
    const server = this.servers.get(serverId);
    if (!server) {
      throw new Error(`MCP server not found: ${serverId}`);
    }

    if (this.clients.has(serverId)) {
      const client = this.clients.get(serverId)!;
      if (client.isConnected()) {
        return; // Already connected
      }
      client.disconnect();
    }

    const client = new MCPClient(server);
    this.clients.set(serverId, client);

    try {
      await client.connect();
      console.log(`Connected to MCP server: ${server.name}`);
    } catch (error) {
      console.error(`Failed to connect to MCP server ${server.name}:`, error);
      this.clients.delete(serverId);
      throw error;
    }
  }

  disconnectServer(serverId: string): void {
    const client = this.clients.get(serverId);
    if (client) {
      client.disconnect();
      this.clients.delete(serverId);
    }

    const server = this.servers.get(serverId);
    if (server) {
      server.status = 'disconnected';
    }
  }

  removeServer(serverId: string): boolean {
    this.disconnectServer(serverId);
    return this.servers.delete(serverId);
  }

  getServer(serverId: string): MCPServer | undefined {
    return this.servers.get(serverId);
  }

  getAllServers(): MCPServer[] {
    return Array.from(this.servers.values());
  }

  getConnectedServers(): MCPServer[] {
    return this.getAllServers().filter(server => server.status === 'connected');
  }

  async callTool(serverId: string, toolName: string, arguments_: Record<string, any>): Promise<any> {
    const client = this.clients.get(serverId);
    if (!client || !client.isConnected()) {
      throw new Error(`MCP server not connected: ${serverId}`);
    }

    return client.callTool(toolName, arguments_);
  }

  getAvailableTools(): Array<{ server: MCPServer; tool: MCPTool }> {
    const tools: Array<{ server: MCPServer; tool: MCPTool }> = [];
    
    for (const server of this.getConnectedServers()) {
      if (server.tools) {
        for (const tool of server.tools) {
          tools.push({ server, tool });
        }
      }
    }

    return tools;
  }
}

export const mcpManager = new MCPManager();

File: server/mcp_client.py
"""MCP Client implementation for LiveKit voice agent"""

import asyncio
import logging
import json
import subprocess
import websockets
from typing import Dict, List, Any, Optional
from datetime import datetime

logger = logging.getLogger("mcp_client")

class MCPServer:
    """Base class for MCP server connections"""
    
    def __init__(self, name: str, params: Dict[str, Any]):
        self.name = name
        self.params = params
        self.server = None
        self.tools = []
        self.connected = False
        
    async def connect(self) -> bool:
        """Connect to the MCP server"""
        raise NotImplementedError
        
    async def disconnect(self):
        """Disconnect from the MCP server"""
        self.connected = False
        
    async def list_tools(self) -> List[Dict[str, Any]]:
        """List available tools from the server"""
        return self.tools
        
    async def call_tool(self, tool_name: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """Call a tool on the MCP server"""
        raise NotImplementedError

class MCPServerStdio(MCPServer):
    """MCP server using stdio protocol (for local servers)"""
    
    def __init__(self, name: str, params: Dict[str, Any]):
        super().__init__(name, params)
        self.process = None
        
    async def connect(self) -> bool:
        try:
            cmd = self.params.get("cmd", "npx")
            args = self.params.get("args", [])
            
            # Start the MCP server process
            self.process = await asyncio.create_subprocess_exec(
                cmd, *args,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            
            # Initialize MCP session
            init_message = {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "initialize",
                "params": {
                    "protocolVersion": "2024-11-05",
                    "capabilities": {
                        "tools": {}
                    },
                    "clientInfo": {
                        "name": "livekit-agent",
                        "version": "1.0.0"
                    }
                }
            }
            
            await self._send_message(init_message)
            response = await self._receive_message()
            
            if response and "result" in response:
                # Get available tools
                tools_message = {
                    "jsonrpc": "2.0", 
                    "id": 2,
                    "method": "tools/list"
                }
                await self._send_message(tools_message)
                tools_response = await self._receive_message()
                
                if tools_response and "result" in tools_response:
                    self.tools = tools_response["result"].get("tools", [])
                
                self.connected = True
                logger.info(f"Connected to {self.name} via stdio, found {len(self.tools)} tools")
                return True
                
        except Exception as e:
            logger.error(f"Failed to connect to {self.name}: {e}")
            await self.disconnect()
            
        return False
    
    async def disconnect(self):
        await super().disconnect()
        if self.process:
            self.process.terminate()
            try:
                await asyncio.wait_for(self.process.wait(), timeout=5.0)
            except asyncio.TimeoutError:
                self.process.kill()
            self.process = None
    
    async def call_tool(self, tool_name: str, params: Dict[str, Any]) -> Dict[str, Any]:
        if not self.connected:
            raise Exception("Server not connected")
            
        message = {
            "jsonrpc": "2.0",
            "id": 3,
            "method": "tools/call",
            "params": {
                "name": tool_name,
                "arguments": params
            }
        }
        
        await self._send_message(message)
        response = await self._receive_message()
        
        if response and "result" in response:
            return response["result"]
        elif response and "error" in response:
            raise Exception(f"Tool call failed: {response['error']}")
        else:
            raise Exception("Invalid response from MCP server")
    
    async def _send_message(self, message: Dict):
        if self.process and self.process.stdin:
            data = json.dumps(message) + "\n"
            self.process.stdin.write(data.encode())
            await self.process.stdin.drain()
    
    async def _receive_message(self) -> Optional[Dict]:
        if self.process and self.process.stdout:
            try:
                line = await asyncio.wait_for(self.process.stdout.readline(), timeout=10.0)
                if line:
                    return json.loads(line.decode().strip())
            except (asyncio.TimeoutError, json.JSONDecodeError) as e:
                logger.error(f"Error receiving message: {e}")
        return None

class MCPServerHttp(MCPServer):
    """MCP server using HTTP/WebSocket protocol (for remote servers)"""
    
    def __init__(self, name: str, params: Dict[str, Any]):
        super().__init__(name, params)
        self.websocket = None
        
    async def connect(self) -> bool:
        try:
            url = self.params.get("url")
            headers = {}
            
            if self.params.get("api_key"):
                headers["Authorization"] = f"Bearer {self.params['api_key']}"
            
            # Connect via WebSocket
            self.websocket = await websockets.connect(
                url.replace("http", "ws"),
                extra_headers=headers
            )
            
            # Initialize MCP session
            init_message = {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "initialize",
                "params": {
                    "protocolVersion": "2024-11-05",
                    "capabilities": {"tools": {}},
                    "clientInfo": {"name": "livekit-agent", "version": "1.0.0"}
                }
            }
            
            await self.websocket.send(json.dumps(init_message))
            response = await self.websocket.recv()
            response_data = json.loads(response)
            
            if "result" in response_data:
                # Get available tools
                tools_message = {"jsonrpc": "2.0", "id": 2, "method": "tools/list"}
                await self.websocket.send(json.dumps(tools_message))
                tools_response = await self.websocket.recv()
                tools_data = json.loads(tools_response)
                
                if "result" in tools_data:
                    self.tools = tools_data["result"].get("tools", [])
                
                self.connected = True
                logger.info(f"Connected to {self.name} via WebSocket, found {len(self.tools)} tools")
                return True
                
        except Exception as e:
            logger.error(f"Failed to connect to {self.name}: {e}")
            await self.disconnect()
            
        return False
    
    async def disconnect(self):
        super().disconnect()
        if self.websocket:
            await self.websocket.close()
            self.websocket = None
    
    async def call_tool(self, tool_name: str, params: Dict[str, Any]) -> Dict[str, Any]:
        if not self.connected or not self.websocket:
            raise Exception("Server not connected")
            
        message = {
            "jsonrpc": "2.0",
            "id": 3,
            "method": "tools/call",
            "params": {"name": tool_name, "arguments": params}
        }
        
        await self.websocket.send(json.dumps(message))
        response = await self.websocket.recv()
        response_data = json.loads(response)
        
        if "result" in response_data:
            return response_data["result"]
        elif "error" in response_data:
            raise Exception(f"Tool call failed: {response_data['error']}")
        else:
            raise Exception("Invalid response from MCP server")